src/engine/Join.cpp

```diff
@@ -746,3 +733,107 @@ void Join::doGallopInnerJoin(const TagType, const IdTableView<L_WIDTH>& l1,
     }
   }
 }
+
+// ______________________________________________________________________________
+template <int L_WIDTH, int R_WIDTH, int OUT_WIDTH>
+void Join::hashJoin(const IdTable& dynA, size_t jc1, const IdTable& dynB,
+                size_t jc2, IdTable* dynRes) {
+  const IdTableView<L_WIDTH> a = dynA.asStaticView<L_WIDTH>();
+  const IdTableView<R_WIDTH> b = dynB.asStaticView<R_WIDTH>();
+
+  LOG(DEBUG) << "Performing hashJoin between two tables.\n";
+  LOG(DEBUG) << "A: width = " << a.numColumns() << ", size = " << a.size() << "\n";
+  LOG(DEBUG) << "B: width = " << b.numColumns() << ", size = " << b.size() << "\n";
+
+  // Check trivial case.
+  if (a.size() == 0 || b.size() == 0) {
+    return;
+  }
+
+  IdTableStatic<OUT_WIDTH> result = std::move(*dynRes).toStatic<OUT_WIDTH>();
+  
+  // Puts the rows of the given table into a hash map, with the value of
+  // the join column of a row as the key, and returns the hash map.
+  auto idTableToHashMap = []<typename Table>(const Table& table, const size_t jc) {
+    // This declaration works, because generic lambdas are just syntactic sugar
+    // for templates. 
+    ad_utility::HashMap<Id, std::vector<typename Table::row_type>> map;
+    for (const auto& row : table) {
+      map[row[jc]].push_back(row);
+    }
+    return map;
+  };
+
+  // Cannot just switch a and b around because the order of
+  // items in the result tuples is important.
+  // Procceding with the actual hash join depended on which IdTableView
+  // is bigger.
+  if (a.size() >= b.size()) {
+    // a is bigger, or the same size as b. b gets put into the hash table.
+    auto map = idTableToHashMap(b, jc2);
+
+
+    // Create cross product by going through a.
+    for (size_t i = 0; i < a.size(); i++) {
+      // Skip, if there is no matching entry in the join column.
+      auto entry = map.find(a(i, jc1));
+      if (entry == map.end()) {
+        continue;
+      }
+
+      for (const auto& bRow: entry->second) {
+        addCombinedRowToIdTable(a[i], bRow, jc2, &result); 
+      }
+    }
```

Comment:
We should still get rid of this code duplication.
I suggest a lambda that takes
```
auto performHashJoin = []<bool rightIsLarger>(largerTable, largerTableJoinColumn, smallerTable, smallerTableJoinColumn) {...
// Everything is the same except for
if constexpr(rightIsLarger) {
   // call the correct addCombinedRow for this case
else {
  // call the other one.
}
}.
```
The templated bool forces the compiler to create the most efficient code, but you then the call is a little bit more complicated:

``` 
if (leftTable < rightTable) {
  performHashJoin.operator()<false>(....);
else {
  performHashJoin.operator()<true>(...);
}
```
and life is good.

Reply:
A good thing is that we have unit tests, so we immediately figure out if something is wrong.

Reply:
I took me a while to find out, but `performHashJoin.operator()<false>(....)` is **not** correct syntax and doesn't compile. You need to write `performHashJoin.template operator()<false>(....)`.

Don't ask me why, I didn't quite understand it either.

Reply:
Done, see [172fc23](https://github.com/ad-freiburg/qlever/pull/821/commits/172fc23b6409d7d928310898c7a4107e89b6524e).